### PR TITLE
EVA-25 Add AggregatedVcfReader and AggregatedVcfLineMapper

### PR DESCRIPTION
**Pull request by jmmut**
Complete spring-based VcfReaders. Add support for aggregated VCFs

### CODE REVIEW

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * {@link AggregatedVcfReader}
+ * input: a Vcf file
+ * output: a list of variants each time its `.read()` is called
+ */
+public class AggregatedVcfReaderTest {
+
+    @Rule
+    public ExpectedException exception = ExpectedException.none();
```
Comment (jorizci) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java]: You are not using it for anything?

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Test
+    public void shouldReadAllLines() throws Exception {
+        ExecutionContext executionContext = MetaDataInstanceFactory.createStepExecution().getExecutionContext();
+
+        // input vcf
+        final String inputFilePath = "/aggregated.vcf.gz";
```
Comment (jorizci) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java]: This should be a private static final String at class level.

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        String input = AggregatedVcfReaderTest.class.getResource(inputFilePath).getFile();
```
Comment (jorizci) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java]: New utils method, TestFileUtils.getResource(...)

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java [hunk 4]
```diff
@@ -105,30 +108,41 @@ public void testConversion() throws Exception {
...
-            for (String innerField : innerFields) {
-                assertNotNull(((BasicDBObject) element).get(innerField));
-                assertFalse(((BasicDBObject) element).get(innerField).toString().isEmpty());
-            }
+        String fileId = "5";
+        String studyId = "7";
+        String studyName = "study name";
+        VcfHeaderReader headerReader = new VcfHeaderReader(new File(inputFile), fileId, studyId, studyName,
+                                                           VariantStudy.StudyType.COLLECTION,
+                                                           VariantSource.Aggregation.NONE);
```
Comment (jorizci) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java]: More repeated initialization code with constants...

File: src/main/java/uk/ac/ebi/eva/pipeline/io/mappers/AggregatedVcfLineMapper.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Maps a String (in VCF format, with aggregated samples) to a list of variants.
+ * <p>
+ * The actual implementation is reused from {@link VariantVcfFactory}.
+ */
+public class AggregatedVcfLineMapper implements LineMapper<List<Variant>> {
+    private VariantSource source;
+
+    private VariantVcfFactory factory;
```
Comment (jorizci) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/mappers/AggregatedVcfLineMapper.java]: Final?

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java (hunk 1 above)
Comment (jmmut) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java]: fixed in next commit

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java (hunk 2 above)
Comment (jmmut) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java]: fixed in next commit

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+        String fileId = "5";
+        String studyId = "7";
+        String studyName = "study name";
+        VcfHeaderReader headerReader = new VcfHeaderReader(new File(input), fileId, studyId, studyName,
+                                                           VariantStudy.StudyType.COLLECTION,
+                                                           VariantSource.Aggregation.NONE);
```
Comment (jmmut) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java]: fixed in next commit

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java (hunk 3 above)
Comment (jmmut) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/AggregatedVcfReaderTest.java]: fixed in next commit

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java (hunk 4 above)
Comment (jmmut) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java]: fixed in next commit

File: src/main/java/uk/ac/ebi/eva/pipeline/io/mappers/AggregatedVcfLineMapper.java (hunk 5 above)
Comment (jmmut) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/mappers/AggregatedVcfLineMapper.java]: fixed in next commit

File: src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java [hunk 7]
```diff
@@ -105,30 +108,41 @@ public void testConversion() throws Exception {
...
+            metadataMongo.append(key, JSON.parse(jsonString));
         }
+
+        checkFieldsInsideList(metadataMongo, "INFO", Arrays.asList("id", "description", "number", "type"));
+        checkStringInsideList(metadataMongo, "contig");
+
+        // the current aggregated.vcf.gz doesn't have FORMAT, FILTER or ALT tags
+//        checkFieldsInsideList(metadataMongo, "FILTER", Arrays.asList("id", "description"));
+//        checkFieldsInsideList(metadataMongo, "ALT", Arrays.asList("id", "description"));
+//        checkFieldsInsideList(metadataMongo, "FORMAT", Arrays.asList("id", "description", "number", "type"));
```
Comment (jmmut) [inline src/test/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReaderTest.java]: fixed in next commit

**Pull request merged**